### ISSUE TITLE
Shader file not found error in Asset Processor

### ISSUE THREAD

**Issue by NiklasHenricson** [o3de/o3de#11893]
**Describe the bug**

This probably needs attention from the core engine group, or whoever deals with the Asset Processor builders. This is not really a sig-graphics issue, since it has nothing to do with rendering shaders. Instead it's an issue about the file structure within external projects that are located outside the C:\o3de structure.

Again, following the same guide under: https://www.o3de.org/docs/learning-guide/tutorials/rendering/get-started-materialtypes-and-shaders/ and it doesn't work. This issue has persisted without a fix since, July 4th, https://github.com/o3de/o3de/issues/10517

The shader cannot be found,

![image](https://user-images.githubusercontent.com/5524755/189510943-4ec806ef-ca78-4f02-9b61-5044722a6c02.png)

While opening Asset Processor's Assets tab it displays my project's file structure and the newly discovered *.materialtype file. It also discovers Atom engine's external references (the reference icons are also clickable and take you to the Atom folder structure). But for the local *.shader file it's not recognized. This is how the JSON file is specified,

```json
    "shaders": 
    [
        {
            "file": "Shaders/Depth/DepthPass.shader"
        },
        {
            "file": "Shaders/Shadow/Shadowmap.shader"
        },
        {
            "file": "Niklas.shader"
        }
    ]	
```

![image](https://user-images.githubusercontent.com/5524755/189511074-a335cdcc-19de-469a-aa9c-e19ee99b4f9a.png)

Tried with the most recent development branch code here,

![image](https://user-images.githubusercontent.com/5524755/189510969-4d74b8b2-0473-478b-a9af-27b2313046a1.png)

**Comment by santorac** [comment]
1. Looking at the code, it appears there should be another error message preceding the one shown in the screenshot. It should say something like "Could not load ... [Source='...' Cache='...' AssetID=...]". Can you confirm whether this is the case?
2. In your screenshot, a couple lines down from Niklas.materialtype there is an error for Niklas.shader. Can you please show us what those error messages are?
3. It might be helpful if attach the actual data files you have that are failing (but I'll try to see if I can repro using my own data).

**Comment by santorac** [comment]
I see the path you are using is "Gems/NavariaOcean/Shaders/...". Usually gem assets go in an assets folder like "Gems/NavariaOcean/**Assets**/Shaders/...". Could this be the cause of the issue?

**Comment by santorac** [comment]
I see you have a "Gems" folder inside your external project. How did you set that up? I've never seen a gems folder under a project folder before.

**Comment by NiklasHenricson** [comment]
I moved practically the material, materialtype, azsl, and shader files across in all directories, restarted O3DE several times, emptying the cache folder in between and waiting for it to re-process the files.

The same outcome whether I place these 4 files in the following structures,

C:\MyProject\Shaders
C:\MyProject\Gems\MyGem\Shaders
C:\MyProject\ShaderLib
C:\MyProject\

Getting the same exact error. The funny thing is, it always finds the Atom-gem shader files referred within the materialtype JSON but nothing else project-related.

I will check tonight about the preceding errors and send you everything that is coming back from the Asset Processor.

**Comment by santorac** [comment]
Okay thanks. 

I suspect what is happening is there's some problem inside the .shader file that fails to build the shader asset, and therefore the .materialtype can't find the shader asset. I have seen users confused by this kind of thing before, where they focus on the wrong problem: some asset A's failure, when the real problem is some other dependency asset B failed first.

I'm not sure it's exactly what you are seeing without more info (we can confirm when  you get a chance to check), but I'll continue working under this assumption...

I was able to get into a similar state, here's what I did....

1. Created a new project "ForNiklas" using the o3de project manager.
2. Created a new gem inside the project using "scripts\o3de.bat create-gem -gp D:\O3DE-Projects\ForNiklas\Gems\NavariaOcean"
3. Copied MinimalPbr.materialtype/shader/azsli from Gems/Atom/TestData/TestData/Materials/Types into ForNiklas/Shaders.
4. Renamed MinimalPbr to "Santorac"
5. Edited Santorac.shader to use badly-formed JSON. This causes the AP to report a CreateJobs failure for that file, similar to what's shown in the original screenshot.

The actual problem is here....
![image](https://user-images.githubusercontent.com/55155825/189734102-90cd8678-3a60-4b4a-9b3b-cceec2fcce1a.png)

But what we see first is this...
![image](https://user-images.githubusercontent.com/55155825/189734185-26021e65-73a5-417f-9191-0b39ef8df63e.png)

Now this log does have enough clues to lead us to the .shader file...
1. This correctly points out that an asset was not processed: "GetAsset called for asset which does not exist in asset catalog and cannot be loaded. Asset may be missing, not processed or moved. AssetId: {FF3A00F8-ABDF-5A53-B95E-79864AB9CBCD}:0". 
2. This identifies the name of the source file that was not processed: "Could not load ShaderAsset [Source='D:\O3DE-Projects\ForNiklas\Gems\NavariaOcean\Shaders\Santorac.shader' Cache='<unknown>' AssetID={FF3A00F8-ABDF-5A53-B95E-79864AB9CBCD}:0]" 
3. The last message "Shader './Santorac.shader' not found" is not exactly correct. The .shader file is found, what wasn't found is the shader asset that should have been be produced by Santorac.shader.

So if I'm correct in all my assumptions, what we need to do here is improve the messaging so the error messages for the .materialtype will clearly lead the user to look at the errors for the .shader. 

@NiklasHenricson do you have any suggestions for error messages that would have helped you know to look at the .shader file problems?

Note, we cannot consolidate these three messages because the come from different areas of code with varying responsibilities.. Message 1 is for all assets, message 2 is for render assets, and message 3 is for material type assets only. We could alter or expand any of these messages to provide more clarity. 

Here's what I suggest for each message...
1. Instead of saying "Asset may be missing, not processed or moved", I'll see if there's a way to detect the actual status from the AP and provide a more specific message like "Asset is missing because asset processing failed" if that's the case.
2. No changes. I can't think of a way to improve this.
3. Update the message to say "Shader asset not found for source file './Santorac.shader'. See above for details."

**Comment by NiklasHenricson** [comment]
@santorac thank you,

You reproduced the exact same error I get. So your conclusion is right. The *.shader file is most likely badly written. I will attach it here in a few hours to provide all the details.

My suggestion for better message would be, 

`Error parsing such-and-such.shader file. The source code line generates a nonspecific syntax error`  and an additional column between "Source" and "Completed" could give us "Error Type/Level" where we apply those three areas you mentioned (Generic Asset, Rendering Asset, Material Type). Or some better naming convention to distunguish those three separate levels. 

Or would it be possible to attach some kind of stack trace from the underlying parsing engine to the error log that is more informative?

I'll provide tonight the source files and the error logs. But your screenshot on the error pretty much nailed my issue. Might not be the same exact syntax issue, but at least it matches with my errors.

**Comment by santorac** [comment]
I don't know about adding a new column for error type. That would be a feature request for sig-core to consider.

Here is the exact error message for my parse error, it's already pretty specific. You just need to know to look at the log for the .shader instead of the log for the .materialtype.
```
>	Failed to parse Shader Descriptor JSON: Failed to load 'D:\O3DE-Projects\ForNiklas\Gems\NavariaOcean\Shaders\Santorac.shader'. JSON parse error at line 1: Invalid value.
```

**Comment by CSteel22** [comment]
I went to check into this for Niklas and went through the process of making my own shader.   I did notice that there were errors farther down in the .shader log that the parser was reporting regarding the following lines (now commented out):
![image](https://user-images.githubusercontent.com/35277774/189767571-ebcf02e5-e898-475b-9081-d9344db8d93f.png)

Once the lines above were commented out, I stumbled upon another a bug.  It appears that if Line 1 of the .azsl file contains a directive, it ignores it.  So when I had this code: 

![image](https://user-images.githubusercontent.com/35277774/189768187-617e978b-a084-49f4-8e73-76c2fc869573.png)

I would get the following errors in the .azsl file:

![image](https://user-images.githubusercontent.com/35277774/189768763-a34c11af-b46b-475c-8e2a-34a22c2eb61c.png)

[Complete log of azsl error when first line contains directive](https://github.com/o3de/o3de/files/9552148/azsl.error.when.first.line.contains.directive.log)

**Comment by santorac** [comment]
Oh that's interesting, @CSteel22 can you please report a bug in the AZSLc repo? https://github.com/o3de/o3de-azslc

**Comment by CSteel22** [comment]
I piggybacked on this issue that was generated back in June: https://github.com/o3de/o3de-azslc/issues/39

**Comment by NiklasHenricson** [comment]
Alright, verified on my end. @CSteel22 description is reproducible and accurate. I can confirm that the .shader file is only compiling when you have a newline on the top of the source file.

**Comment by santorac** [comment]
> We may be able to detect that the source file failed processing to provide a more specific error message like this: "Could not load ShaderAsset because the source file 'D:\O3DE-Projects\ForNiklas\Gems\NavariaOcean\Shaders\Santorac.shader' failed to process. See Santorac.shader's log for details."

I will not be able to improve this error message with the current system, I opened a feature request to unblock it: https://github.com/o3de/o3de/issues/12018

> Update the message to say "Shader asset not found for source file './Santorac.shader'. See above for details."

This is all I can do for now.

**Comment by NiklasHenricson** [comment]
Thanks @santorac I appreciate your help. In meanwhile we'll live with the workaround. Based on that knowledge alone we've made some progress and moved on for now.

**Comment by tonybalandiuk** [comment]
@santorac based on the timing of this being created, I think it likely exists in stabilization. If so, do you think this needs to be resolved in the upcoming release? If so, we should add it to the "Release/2210" milestone. If not, please add it to the "Backlogged Issues" milestone. Thanks. @AMZN-Dk  and I are working to understand what's still required to fix for the upcoming release - your assistance here will help us in that endeavor. 
Also do you think this is something that should be noted as a "known issue"?

**Comment by santorac** [comment]
@tonybalandiuk IMO this does not need to be addressed for the upcoming release. It's just about improving error messaging a bit, but the current messaging is sufficient for now.

https://github.com/o3de/o3de-azslc/issues/39 might deserve consideration for the release though.

**Comment by santorac** [comment]
I'm closing this due to a combination of:

1. Workaround guidance provided.
2. Error message improved in this PR https://github.com/o3de/o3de/pull/12019
3. New ticket open for further error message improvement: https://github.com/o3de/o3de/issues/12018
